spec/index.bs

```diff
@@ -4242,6 +4242,11 @@ GPUCommandEncoder includes GPUObjectBase;
     : <dfn>\[[debug_group_stack]]</dfn> of type [=stack=]&lt;{{USVString}}&gt;.
     ::
         A stack of active debug group labels.
+
+    : <dfn>\[[queries_state_map]]</dfn>, of type [=ordered map=]&lt;{{GPUQuerySet}}, sequence&lt;{{boolean}}&gt;&gt;
+    ::
+        The queries state for each {{GPUQuerySet}} indicating which queries are used on the {{GPUCommandEncoder}},
+        initially empty.
```

Comment:
I find the sequence of boolean is not enough for representing the query state in occlusion query and pipeline statistics query, because except the used query index, we also need to know whether a query is begun and ended. I will update this with a enum of query state instead as below:
Unavailable - The query is unused.
Active - The query is using but not finished writing.
Available - The query is finished and ready for resolving.

Reply:
I'm not sure this is the correct solution but I'm unclear on the problems we are trying to solve. Is this the correct list of constraints we're trying to encode?

 - The same query cannot be written to twice in the same render pass because of synchronization issues.
 - Queries that are started on an encoder must be ended before the call to finish on that encoder.
 - Queries that are started cannot be started again before they are ended.

The biggest constraints are on the render pass because we can't introduce barriers inside it, but outside of render passes I think it should be valid to do the following:

```
const computeEncoder = commandEncoder.begingComputePass();
computeEncoder.writeTimestamp(timestampSet, 0);
computeEncoder.writeTimestamp(timestampSet, 0);
computeEncoder.beginPipelineStastistics(statSet, 0);
computeEncoder.endPipelineStastistics(statSet, 0);
computeEncoder.beginPipelineStastistics(statSet, 0);
computeEncoder.endPipelineStastistics(statSet, 0);
```

Reply:
Also FYI the tracking we'll need in WebGPU implementation will be more complex than what's described in the specification because we need to end / reset queries, track ones uninitialized so we can produces 0s, etc. But this tracking shouldn't appear in the WebGPU specification because it is an implementation detail.

Reply:
Yes, the biggest constraints are on the render pass, but I expect to have the same restrictions for a same method on both command encoder and render/compute encoders. If it allows to overwrite  a same query on command encoder and compute encoder, but does not allow on render encoder, I don't think it make sense. 

So the restrictions list I proposed is:
- The same query cannot be written to twice on the same command encoder.
- Queries that are started on an encoder must be ended before the call to finish on that encoder.
- Queries that are started cannot be started again before they are ended.
- Queries cannot be ended without corresponding start operation.

If we describe the restrictions as above in the validations, I don't think we need to add the sequence of boolean which play a same role in the validations.

Reply:
> If it allows to overwrite a same query on command encoder and compute encoder, but does not allow on render encoder, I don't think it make sense.

It makes sense for the render encoder to have more restrictions because it is a single synchronization scope. Maybe we can restrict it for compute encoders (even though it's not strictly necessary). However we should be able to use the same query twice in a command encoder because the resolve command is on that encoder.

Reply:
+1, I think it should be possible to write, resolve, write, resolve in a single command buffer. Generally I think command buffers never have restrictions like this (it should generally be possible to use a single command buffer any time you could have used multiple).